### DISCUSSION TITLE
使用 pnpm 代替 yarn + lerna 构建项目

### DISCUSSION THREAD

**Discussion by liuweiGL** [alibaba/formily#1745] [Ideas]
## 当前存在的问题

使用 lerna 进行 build 时效率太低了。

## 原因

lerna 需要先分析包的执行顺序，然后按顺序执行。

## 尝试过的方法

使用 `--parallel` 进行并行构建，但是失败了。因为当前 package 的 main 指向的是构建之后的目录，而不是 src 源码；所以存在依赖的包无法同时进行构建，tsc 找不到依赖会报错。

## pnpm 的解决办法

1. 在 package.json 中把 main/module 字段指向 `src/index.ts`，这样开发的时候包之间直接引用源码。
2. 使用 [publishconfig](https://pnpm.io/zh/package_json#publishconfig) 配置重写入口配置，指向构建目录。
3. 包之间没有构建依赖，可以使用 parallel 模式 build。

**Comment by janryWang** [comment]
本身build就是依赖顺序的。。。因为要构建类型，否则build的时候类型拿不到会报错的

**Reply by liuweiGL** [reply]
把 main 指向 src 就不需要构建依赖了。

**Reply by janryWang** [reply]
package.main肯定改不了呀

**Reply by liuweiGL** [reply]
为什么改不了，pnpm就可以开发一个配置，publish 一个配置啊。

**Reply by janryWang** [reply]
main指向src，那依赖库的使用者webpack打包，没有用ts就挂了

**Comment by liuweiGL** [comment]
等我本地测试下 pnpm 的 publishConfig 效果。

**Comment by liuweiGL** [comment]
我弄了个 demo，应该是可行的：https://github.com/liuweiGL/pnpm-workspace

**Comment by janryWang** [comment]
研究了一下pnpm，感觉它思想上倒是没啥问题，问题主要是它不支持隐式依赖吧，比如项目里依赖了A组件，A组件依赖了B组件，那么在项目里没有明确声明B组件依赖的话，是没法使用B组件的，我不确定这个特性到底好不好，但是我能确定肯定有用户项目存在隐式依赖，formily本身估计也有，因为这个完全是受yarn npm影响的潜意识行为，然后就是迁移pnpm之后，需要把formily所有隐式依赖给明确声明掉

还有就是，如何调试依赖的第三方库，比如用到了某个第三方库，因为它把依赖全部去重放到了磁盘根目录上，那么想要快速修改第三方库源码来调试就很难调试了，如果强制跑到根目录上去修改第三方库源码，风险很高，因为会影响其他仓库

vscode体验，能基于第三方库的类型签名跳到指定位置吗？这个不太确定，如果不能，那体验太糟糕

发布控制，它能否做到private包不发布？

**Comment by liuweiGL** [comment]
> 问题主要是它不支持隐式依赖吧，比如项目里依赖了A组件，A组件依赖了B组件，那么在项目里没有明确声明B组件依赖的话，是没法使用B组件的，我不确定这个特性到底好不好，但是我能确定肯定有用户项目存在隐式依赖，formily本身估计也有，因为这个完全是受yarn npm影响的潜意识行为，然后就是迁移pnpm之后，需要把formily所有隐式依赖给明确声明掉

1. 隐式依赖肯定是不对的吧，版本控制权都不在自己手里，太不可靠了。
2. 用户有没有隐式依赖应该跟我们没关系。
3. formily 存在隐式依赖的，demo 仓库就是调整过的。
4. pnpm 也可以支持依赖提升，而且方式有很多，这个不会是问题。

> 还有就是，如何调试依赖的第三方库，比如用到了某个第三方库，因为它把依赖全部去重放到了磁盘根目录上，那么想要快速修改第三方库源码来调试就很难调试了，如果强制跑到根目录上去修改第三方库源码，风险很高，因为会影响其他仓库

1. 修改是没问题的，就像正常的 npm 包一样就好了。
2. 如果怕修改之后忘记改回去影响其他仓库，可能需要直接拷贝一份然后使用 file:// 协议安装。

> vscode体验，能基于第三方库的类型签名跳到指定位置吗？这个不太确定，如果不能，那体验太糟糕

这个没问题的，不然在前端肯定是用不起来的，我现在所有的项目都切换到 pnpm 了。

> 发布控制，它能否做到private包不发布？

经测试，可以的：
![image](https://user-images.githubusercontent.com/21048878/125012296-8b35a200-e09c-11eb-802e-8345c836ac1e.png)

**Reply by henryzp** [reply]
隐式依赖，在formily v1里面有坑。。。v2里面，测试一下，挺好的

**Comment by MatriXiao88** [comment]
刚好看到这个推送，权当参考吧
[在字节用 Monorepo，我们帮你踩了这些坑](https://mp.weixin.qq.com/s/KExxFGJASD-DczC5BCGNMQ)

**Reply by liuweiGL** [reply]
文章不错呀！
Rush 我之前看过，感觉稍微有点复杂的，目前 pnpm 的 workspace 功能完全够用了。

**Reply by liuweiGL** [reply]
奥，还不是，我之前看的是 Rush Stack 。

**Comment by janryWang** [comment]
可以先提交个PR，我checkout下来实际试用一下看看？

**Reply by janryWang** [reply]
贡献文档得改一下

**Comment by liuweiGL** [comment]
恩，等我后面提个 PR。

**Reply by janryWang** [reply]
prettier/eslint/githook这个很难搞啊

**Reply by liuweiGL** [reply]
啊？这些跟包管理器有关系吗？

**Reply by janryWang** [reply]
就那篇微信文章里讲了呀

**Reply by liuweiGL** [reply]
你是指这些包的依赖会有问题吗？

**Reply by janryWang** [reply]
我的意思是现在prettier/eslint/githook都是全局依赖，切到pnpm了之后没法做全局依赖了呀

**Reply by liuweiGL** [reply]
文章里指的是 Rush 类库存在的问题，我们又不用它，所以这些问题都不是问题。

**Reply by liuweiGL** [reply]
> 我的意思是现在prettier/eslint/githook都是全局依赖，切到pnpm了之后没法做全局依赖了呀

切换到 pnpm 最重要的改变的只是 package.json 中可以使用 publishConfig 来重写 main/module 这些字段配置，这样的话就可以开发的时候 main 指向 src，publish 的时候 main 指向 esm。

其他的东西还是跟现在保持一致的，包括项目源码、依赖目录结构。

**Reply by janryWang** [reply]
嗯嗯，可以试一下，先提个pr的时候我看看

**Comment by janryWang** [comment]
还有个问题，如果用了pnpm，那么全局跑jest，能否做到跨包共享覆盖率，比如我在A包里的tests里引用了B包的API，其实是走到了B包的某个分支，结果覆盖率提不上去，目前走yarn是没法跨包共享覆盖率的

**Reply by liuweiGL** [reply]
这块我不太了解，等我看看吧。
